File: leetcode/pjkfckr/Container_With_Most_Water.py
"""
주어진 정수 배열 height의 길이는 n입니다. 두 개의 수직선이 그려져 있으며, i번째 선의 두 끝점은 (i, 0)과 (i, height[i])입니다.
x축과 함께 컨테이너를 형성하는 두 개의 선을 찾아야 하며, 이 컨테이너가 가장 많은 물을 담을 수 있도록 해야 합니다.
컨테이너가 담을 수 있는 최대 물의 양을 반환하세요.
컨테이너를 기울일 수 없다는 점에 유의하세요.

"""

height = [1, 1]

left, right = 0, len(height) - 1
max_area = 0

while left < right:
    width = right - left
    container_height = min(height[left], height[right])
    area = width * container_height
    max_area = max(max_area, area)

    if height[left] < height[right]:
        left += 1
    else:
        right -= 1

print(max_area)

File: leetcode/0chnxxx/Container_With_Most_Water.kt
/**
 * You are given an integer array height of length n. There are n vertical lines drawn such that the two endpoints of the ith line are (i, 0) and (i, height[i]).
 *
 * Find two lines that together with the x-axis form a container, such that the container contains the most water.
 *
 * Return the maximum amount of water a container can store.
 *
 * Notice that you may not slant the container.
 *
 * n == height.length
 * 2 <= n <= 10^5
 * 0 <= height[i] <= 10^4
 */

fun main() {
    val height = intArrayOf(1, 8, 6, 2, 5, 4, 8, 3, 7)

    val solution = Solution().maxArea(height)

    println(solution)
}

class Solution {
    // O(N) 풀이
    fun maxArea(height: IntArray): Int {
        var max = 0
        var left = 0
        var right = height.size - 1

        while (left < right) {
            val vertical = minOf(height[left], height[right])
            val horizontal = right - left
            val area = vertical * horizontal

            max = maxOf(max, area)

            if (height[left] < height[right]) {
                left++
            } else {
                right--
            }
        }

        return max
    }
}
